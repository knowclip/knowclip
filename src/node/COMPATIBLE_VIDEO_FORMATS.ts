/** List from https://www.chromium.org/audio-video/
 * formats from ffmpeg -formats
 */

import { FfprobeData } from 'fluent-ffmpeg'

export const COMPATIBLE_VIDEO_FORMATS = {
  // MP4 (QuickTime/ MOV / ISO-BMFF / CMAF)
  // Ogg
  // WebM
  // WAV
  // HLS [Only on Android and only single-origin manifests]
  mp4: 'MP4 (MPEG-4 Part 14)',
  ogg: 'Ogg',
  webm: 'WebM',
  ['matroska,webm']: 'Matroska / WebM',
  matroska: 'Matroska',
  wav: 'WAV / WAVE (Waveform Audio)',
}
/** List from https://www.chromium.org/audio-video/
 * codecs from ffmpeg -codecs
 */
export const COMPATIBLE_AUDIO_CODECS = {
  // FLAC
  // MP3
  // Opus
  // PCM 8-bit unsigned integer
  // PCM 16-bit signed integer little endian
  // PCM 32-bit float little endian
  // PCM Î¼-law
  // Vorbis
  // AAC [Main, LC, HE profiles only, xHE-AAC on Android P+, macOS, Windows 11] [Google Chrome only]
  flac: 'FLAC (Free Lossless Audio Codec)',
  mp3: 'MP3 (MPEG audio layer 3)',
  opus: 'Opus',
  pcm_u8: 'PCM unsigned 8-bit',
  pcm_s16le: 'PCM signed 16-bit little-endian',
  pcm_f32le: 'PCM 32-bit float little-endian',
  pcm_mulaw: 'PCM mu-law',
  vorbis: 'Vorbis',
  aac: 'AAC (Advanced Audio Coding)',
}
/** List from https://www.chromium.org/audio-video/
 * codecs from ffmpeg -codecs
 */
export const COMPATIBLE_VIDEO_CODECS = {
  // AV1
  // VP8
  // VP9
  // H.264 [Google Chrome only]
  // H.265 [Google Chrome only and only where supported by the underlying OS]
  av1: 'AV1',
  vp8: 'VP8',
  vp9: 'VP9',
  h264: 'H.264',
  h265: 'H.265',
}

export function getVideoCompatibilityWarnings(metadata: FfprobeData) {
  const warnings: string[] = []
  const fileFormat = metadata.format.format_name
  if (!fileFormat || !(fileFormat.toLowerCase() in COMPATIBLE_VIDEO_FORMATS)) {
    warnings.push(`Video format ${fileFormat} is not compatible`)
  }

  const videoStreams = metadata.streams.filter(
    (stream) => stream.codec_type && /video/i.test(stream.codec_type)
  )
  const audioStreams = metadata.streams.filter(
    (stream) => stream.codec_type && /audio/i.test(stream.codec_type)
  )
  // warn if multiple audio or video streams
  if (videoStreams.length > 1) {
    warnings.push(`Multiple video streams detected`)
  }
  if (audioStreams.length > 1) {
    warnings.push(`Multiple audio streams detected`)
  }

  const incompatibleVideoCodecs = videoStreams.filter((stream) => {
    const codecName = stream.codec_name
    return !(codecName && codecName.toLowerCase() in COMPATIBLE_VIDEO_CODECS)
  })
  if (incompatibleVideoCodecs.length) {
    warnings.push(
      `Incompatible video codecs: ${incompatibleVideoCodecs
        .map((stream) => stream.codec_name)
        .join(', ')}`
    )
  }

  const incompatibleAudioCodecs = audioStreams.filter((stream) => {
    const codecName = stream.codec_name
    return !(codecName && codecName.toLowerCase() in COMPATIBLE_AUDIO_CODECS)
  })
  if (incompatibleAudioCodecs.length) {
    warnings.push(
      `Incompatible audio codecs: ${incompatibleAudioCodecs
        .map((stream) => stream.codec_name)
        .join(', ')}`
    )
  }

  return warnings
}
