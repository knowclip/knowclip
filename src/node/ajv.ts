import Ajv from 'ajv'
import betterAjvErrors from 'better-ajv-errors'

const ajv = new Ajv()

export const compile = ajv.compile.bind(ajv)

export { betterAjvErrors }
