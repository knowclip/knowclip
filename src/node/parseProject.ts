import { parse } from 'yaml'
import { readFile } from 'fs/promises'
import { ProjectJson } from '../types/Project'
import validateProject from './validateProject'

export type NormalizedProjectFileData = {
  project: ProjectFile
  media: MediaFile[]
  clips: Clip[]
  cards: Flashcard[]
  subtitles: SubtitlesFile[]
}

export const parseProjectJson = async <F extends FlashcardFields>(
  filePath: string
): AsyncResult<ProjectJson<F>> => {
  try {
    const docsText = (await readFile(filePath, 'utf8'))
      .split(/(?:^|\r?\n)(?:---|\.\.\.)\r?\n/)
      .filter((x) => x.trim())
    const errors: string[] = []
    const docs = docsText.map((text) => {
      try {
        const doc = parse(text, {
          maxAliasCount: -1,
        })
        return doc
      } catch (err) {
        console.error(err)
        errors.push(String(err))
        return null
      }
    })

    if (errors.length) return { errors }

    const [project, ...media] = docs

    const validation = validateProject(project, media)

    if (validation.errors)
      return {
        errors: Object.entries(validation.errors).map(
          ([sectionName, bigErrorMessage]) => {
            return `Invalid data for ${sectionName}:\n\n${bigErrorMessage}`
          }
        ),
      }

    return {
      value: {
        project,
        media,
      },
    }
  } catch (err) {
    return { errors: [String(err)] }
  }
}
