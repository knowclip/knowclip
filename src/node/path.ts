export { basename, extname, dirname, resolve, join } from 'path'
