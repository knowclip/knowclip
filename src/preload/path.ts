import { getPreloadModule } from './getPreloadModule'

const { basename, extname, dirname, resolve, join } =
  getPreloadModule('path') || require('path')

export { basename, extname, dirname, resolve, join }
