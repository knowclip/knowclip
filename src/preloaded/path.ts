import { getPreloadModule } from './getPreloadModule'

const { basename, extname, dirname, resolve, join } = getPreloadModule('path')

export { basename, extname, dirname, resolve, join }
