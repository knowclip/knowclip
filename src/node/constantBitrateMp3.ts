import { createConstantBitrateMp3 } from './ffmpeg'
import { existsSync } from 'fs'
import { extname, basename, join } from 'path'
import { rootTemporaryDirectory } from 'tempy'

export const coerceMp3ToConstantBitrate = (
  path: string,
  oldConstantBitratePath: string | null
): Promise<string> => {
  // should check if mp3
  // and if possible, constant vs. variable bitrate
  return new Promise((res, rej) => {
    if (extname(path) !== '.mp3') return res(path)
    if (oldConstantBitratePath && existsSync(oldConstantBitratePath))
      return res(oldConstantBitratePath)

    const constantBitratePath = join(
      rootTemporaryDirectory,
      basename(path, '.mp3') + '.mp4'
    )

    // I guess by default it does CBR
    // though maybe we should check that
    // bitrate and buffersize defaults are ok.
    //   .outputOptions('-bufsize 192k')
    return createConstantBitrateMp3(path, res, constantBitratePath, rej)
  })
}
