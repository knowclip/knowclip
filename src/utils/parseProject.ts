import { parse } from 'yaml'
import { readFile } from 'preloaded/fs'
import { blankSimpleFields, blankTransliterationFields } from './newFlashcard'
import { parseFormattedDuration } from './formatTime'
import {
  ProjectJson,
  MediaJson,
  SubtitlesJson,
  EmbeddedSubtitlesJson,
} from '../types/Project'
import validateProject from './validateProject'
import { unescapeClozeFields } from './clozeField'

type NormalizedProjectFileData = {
  project: ProjectFile
  media: MediaFile[]
  clips: Clip[]
  cards: Flashcard[]
  subtitles: SubtitlesFile[]
}

export const parseProjectJson = async <F extends FlashcardFields>(
  filePath: string
): AsyncResult<ProjectJson<F>> => {
  try {
    console.log('reading', filePath)
    const docsText = (await readFile(filePath))
      .split(/(?:^|\n)(?:---|\.\.\.)\n/)
      .filter((x) => x.trim())
    const errors: string[] = []
    console.log('parsing', docsText)
    const docs = docsText.map((text) => {
      try {
        const doc = parse(text, {
          maxAliasCount: -1,
        })
        return doc
      } catch (err) {
        console.error(err)
        console.log('error:::', { text })
        errors.push(String(err))
        return null
      }
    })

    if (errors.length) return { errors }

    const [project, ...media] = docs

    const validation = validateProject(project, media)

    if (validation.errors)
      return {
        errors: Object.entries(validation.errors).map(
          ([sectionName, bigErrorMessage]) => {
            return `Invalid data for ${sectionName}:\n\n${bigErrorMessage}`
          }
        ),
      }

    return {
      value: {
        project,
        media,
      },
    }
  } catch (err) {
    return { errors: [String(err)] }
  }
}

export const normalizeProjectJson = <F extends FlashcardFields>(
  state: AppState,
  { project: projectJson, media: mediaJson }: ProjectJson<F>
): NormalizedProjectFileData => {
  const project: ProjectFile = {
    id: projectJson.id,
    createdAt: projectJson.createdAt,
    lastSaved: projectJson.timestamp,
    noteType: projectJson.noteType,
    name: projectJson.name,
    mediaFileIds: mediaJson.map((m) => m.id),
    type: 'ProjectFile',
    error: null,
  }
  const media: [MediaFile, () => Clip[], () => Flashcard[], SubtitlesFile[]][] =
    mediaJson.map((m) => {
      const subtitles = m.subtitles
        ? m.subtitles.map((s) => toMediaSubtitlesRelation(s))
        : []
      const subtitlesFiles: SubtitlesFile[] = (m.subtitles || []).map(
        (s): SubtitlesFile =>
          s.type === 'Embedded'
            ? {
                type: 'VttConvertedSubtitlesFile',
                id: s.id,
                streamIndex: s.streamIndex,
                parentId: m.id,
                parentType: 'MediaFile',
                chunksMetadata: s.chunksMetadata,
              }
            : {
                type: 'ExternalSubtitlesFile',
                id: s.id,
                name: s.name,
                parentId: m.id,
                chunksMetadata: s.chunksMetadata,
              }
      )
      const base: AudioFile = {
        name: m.name,
        type: 'MediaFile',
        id: m.id,
        parentId: projectJson.id,
        durationSeconds: parseFormattedDuration(m.duration).asSeconds(),
        format: m.format,
        subtitles,
        flashcardFieldsToSubtitlesTracks:
          m.flashcardFieldsToSubtitlesTracks || {},
        subtitlesTracksStreamIndexes: (m.subtitles || [])
          .filter((s): s is EmbeddedSubtitlesJson => s.type === 'Embedded')
          .map((s) => s.streamIndex),

        isVideo: false,
      }

      if ('width' in m) {
        return [
          {
            ...base,
            isVideo: true,
            width: m.width,
            height: m.height,
          },
          getMediaClipsGetter<F>(state, project, m),
          getMediaCardsGetter<F>(state, project, m),
          subtitlesFiles,
        ]
      }

      return [
        base,
        getMediaClipsGetter<F>(state, project, m),
        getMediaCardsGetter<F>(state, project, m),
        subtitlesFiles,
      ]
    })

  let clips: Clip[]
  let cards: Flashcard[]

  return {
    project,
    media: media.flatMap(([m]) => m),
    get clips() {
      clips = clips || media.flatMap(([, c]) => c())
      return clips
    },
    get cards() {
      cards = cards || media.flatMap(([, , c]) => c())
      return cards
    },
    subtitles: media.flatMap(([, , , s]) => s),
  }
}

const getMediaClipsGetter =
  <F extends FlashcardFields>(
    state: AppState,
    project: ProjectFile,
    media: MediaJson<F>
  ) =>
  () =>
    getMediaClips(state, project, media)
const getMediaCardsGetter =
  <F extends FlashcardFields>(
    state: AppState,
    project: ProjectFile,
    media: MediaJson<F>
  ) =>
  () =>
    getMediaCards(state, project, media)

function getMediaClips<F extends FlashcardFields>(
  state: AppState,
  project: ProjectFile,
  media: MediaJson<F>
): Clip[] {
  return (media.clips || []).map((c): Clip => {
    return {
      id: c.id,
      clipwaveType: 'Primary',
      start: parseFormattedDuration(c.start).asMilliseconds(),
      end: parseFormattedDuration(c.end).asMilliseconds(),
      fileId: media.id,
    }
  })
}
function getMediaCards<F extends FlashcardFields>(
  state: AppState,
  project: ProjectFile,
  media: MediaJson<F>
): Flashcard[] {
  return (media.clips || []).map((c): Flashcard => {
    const flashcardBase = {
      type: project.noteType,
      id: c.id,
      tags: c.tags || [],
      image: c.image && {
        id: c.id,
        type: 'VideoStillImage' as const,
        ...(c.image.time
          ? { seconds: parseFormattedDuration(c.image.time).asSeconds() }
          : null),
      },
    }

    const { fields, cloze } = unescapeClozeFields(c.fields)

    return project.noteType === 'Simple'
      ? {
          ...flashcardBase,
          type: 'Simple',
          fields: { ...blankSimpleFields, ...fields },
          cloze,
        }
      : {
          ...flashcardBase,
          type: 'Transliteration',
          fields: { ...blankTransliterationFields, ...fields },
          cloze,
        }
  })
}

function toMediaSubtitlesRelation(s: SubtitlesJson): MediaSubtitlesRelation {
  switch (s.type) {
    case 'Embedded':
      return {
        type: 'EmbeddedSubtitlesTrack',
        id: s.id,
      }

    case 'External':
      return {
        type: 'ExternalSubtitlesTrack',
        id: s.id,
      }
  }
}
