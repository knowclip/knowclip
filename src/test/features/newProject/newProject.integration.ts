import {
  initTestContext,
  startApp,
  stopApp,
  IntegrationTestContext,
} from '../../setUpDriver'
import { mockSideEffects } from '../../mockSideEffects'
import { runAll } from '../step'
import { newProjectTestSteps } from './newProjectTestSteps'
import { join } from 'path'
import { parseProjectJson } from '../../../node/parseProject'
import { describe, beforeAll, afterAll, test, expect } from 'vitest'

describe('create a deck from a new project', () => {
  const context: IntegrationTestContext = initTestContext('newProject')

  beforeAll(async () => {
    const { app } = await startApp(context, 9515)

    await mockSideEffects(app, sideEffectsMocks)
  })

  runAll(
    newProjectTestSteps({
      projectFileName: 'my_cool_new_project',
      projectTitle: 'My cool new project',
      projectId: '6f58206b-9384-411c-8432-e7ff8d6c958b',
      mediaFileIds: [
        '5b0db4e9-a28f-4499-9cd7-d775f4eb7cb0',
        '906db000-f3d2-46c9-9ec9-0a1069304e41',
      ],
      clipIds: [
        '8e92ca1d-8d29-486b-81d1-0a44d2da8366',
        'a1eb3142-ce18-4595-a3c3-8c867e1058f4',
        '3eb212dd-8f04-4c63-b36a-06a0bb6957aa',
      ],
      subtitlesTrackIds: [
        '7077af99-722b-4737-854e-d2d1a3a3a60f',
        'a517b52e-e10b-4e9e-90e1-bfecc865b428',
      ],
    }),
    context
  )

  test('resulting project file matches snapshot', async () => {
    const actualProjectFileContents = await parseProjectJson(
      join(context.temporaryDirectory, 'my_cool_new_project.kyml')
    )

    expect(actualProjectFileContents).toMatchSnapshot()
  })

  afterAll(async () => {
    await stopApp(context)
  })
})

const sideEffectsMocks = {
  nowUtcTimestamp: [
    '2020-01-28T12:35:20Z',
    '2020-01-28T12:35:20Z',
    '2020-01-28T12:35:20Z',
    '2020-01-28T12:35:49Z',
    '2020-01-31T14:26:39Z',
    '2020-01-31T14:26:40Z',
    '2020-01-31T14:26:41Z',
    '2020-01-31T14:26:42Z',
    '2020-01-31T14:26:42Z',
    '2020-01-31T14:26:43Z',
    '2020-01-31T14:26:44Z',
    '2020-01-31T14:26:52Z',
    '2020-01-31T14:26:52Z',
    '2020-01-31T14:27:03Z',
    '2020-01-31T14:27:04Z',
    '2020-01-31T14:27:04Z',
    '2020-01-31T14:27:05Z',
    '2020-01-31T14:27:05Z',
    '2020-01-31T14:27:05Z',
    '2020-01-31T14:27:05Z',
    '2020-01-31T14:27:05Z',
    '2020-01-31T14:27:08Z',
    '2020-01-31T14:27:08Z',
    '2020-01-31T14:27:08Z',
    '2020-01-31T14:27:08Z',
    '2020-01-31T14:29:18Z',
    '2020-01-31T14:29:18Z',
    '2020-01-31T14:29:18Z',
    '2020-01-31T14:29:18Z',
    '2020-01-31T14:29:18Z',
    '2020-01-31T14:29:18Z',
    '2020-01-31T14:29:18Z',
    '2020-01-31T14:29:19Z',
    '2020-01-31T14:29:19Z',
    '2020-01-31T14:29:19Z',
    '2020-01-31T14:29:19Z',
  ],
}
