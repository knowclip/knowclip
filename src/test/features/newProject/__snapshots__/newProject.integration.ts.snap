// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`create a deck from a new project > resulting project file matches snapshot 1`] = `
{
  "value": {
    "media": [
      {
        "clips": [
          {
            "end": "00:12.5",
            "fields": {
              "meaning": "Relaxing while eating bamboo grass is the best",
              "pronunciation": "sasa-o tabe-nágara nonbíri-suru-no-wa saikoo-da-naa",
              "transcription": "笹を食べながらのんびりするのは最高だなぁ",
            },
            "id": "8e92ca1d-8d29-486b-81d1-0a44d2da8366",
            "image": {
              "type": "VideoStill",
            },
            "start": "00:08.04",
            "tags": [
              "pbc",
            ],
          },
          {
            "end": "00:17.76",
            "fields": {
              "meaning": "This kid, lazing about again so early",
              "notes": "\\"Goro-goro\\" is the sound of something big rolling around.",
              "pronunciation": "mata kono ko-wa hirumá-kara góro-goro shite",
              "transcription": "またこの子は昼間からゴロゴロして",
            },
            "id": "a1eb3142-ce18-4595-a3c3-8c867e1058f4",
            "image": {
              "type": "VideoStill",
            },
            "start": "00:15.12",
            "tags": [
              "pbc",
            ],
          },
          {
            "end": "01:21.16",
            "id": "3eb212dd-8f04-4c63-b36a-06a0bb6957aa",
            "image": {
              "type": "VideoStill",
            },
            "start": "01:15.2",
            "tags": [
              "pbc",
            ],
          },
        ],
        "duration": "01:27.062",
        "format": "mov,mp4,m4a,3gp,3g2,mj2",
        "id": "5b0db4e9-a28f-4499-9cd7-d775f4eb7cb0",
        "name": "polar_bear_cafe.mp4",
        "subtitles": [
          {
            "chunksMetadata": {
              "count": 26,
              "endCueMs": 86430,
            },
            "id": "7077af99-722b-4737-854e-d2d1a3a3a60f",
            "streamIndex": 2,
            "type": "Embedded",
          },
          {
            "chunksMetadata": {
              "count": 15,
              "endCueMs": 86730,
            },
            "id": "a517b52e-e10b-4e9e-90e1-bfecc865b428",
            "name": "pbc_jp.ass",
            "type": "External",
          },
        ],
      },
      {
        "duration": "00:58.24",
        "format": "mov,mp4,m4a,3gp,3g2,mj2",
        "id": "906db000-f3d2-46c9-9ec9-0a1069304e41",
        "name": "piggeldy_cat.mp4",
      },
    ],
    "project": {
      "createdAt": "2020-01-28T12:35:20Z",
      "id": "6f58206b-9384-411c-8432-e7ff8d6c958b",
      "name": "My cool new project",
      "noteType": "Transliteration",
      "timestamp": "2020-01-31T14:27:05Z",
    },
  },
}
`;
